### ISSUE TITLE
Incorrect CPU implementation of CTCLoss backward step

### ISSUE THREAD

**Issue by wx-csy** [pytorch/pytorch#82545]
### 🐛 Describe the bug

The ATen CTCLoss backward step seems to produce incorrect gradients. The following code snippet reproduces this issue. It computes the differences of the output with respect to every input parameter `input`, and compares them with the backpropagated gradients. The results shall be approximately equal. 

``` python
#!/usr/bin/env python3
import torch

torch.manual_seed(0)
torch.set_printoptions(precision=10)

T, C, S = 5, 3, 4

input = torch.randn(T, C).log_softmax(1).detach().requires_grad_()
target = torch.randint(low=1, high=C, size=(S,), dtype=torch.long)

loss = torch.nn.functional.ctc_loss(
    input, target, torch.tensor(T), torch.tensor(S))
loss.backward()

for i in range(T):
    for j in range(C):
        new_input = input.clone().detach()
        new_input[i][j] += 0.01

        new_loss = torch.nn.functional.ctc_loss(
            new_input, target, torch.tensor(T), torch.tensor(S))

        print((new_loss - loss).detach(), input.grad[i][j] * 0.01) # expected to be approximately equal
```

The actual output is shown below, and the two columns differ significantly.

```
tensor(0.) tensor(0.0021115856)
tensor(0.) tensor(0.0003372314)
tensor(-0.0024995804) tensor(-0.0024488156)
tensor(0.) tensor(0.0018777853)
tensor(-0.0024995804) tensor(-0.0021404340)
tensor(0.) tensor(0.0002626498)
tensor(0.) tensor(0.0008711107)
tensor(0.) tensor(0.0013454026)
tensor(-0.0024995804) tensor(-0.0022165121)
tensor(-0.0025000572) tensor(-0.0018093735)
tensor(0.) tensor(0.0005692409)
tensor(0.) tensor(0.0012401351)
tensor(0.) tensor(0.0002813108)
tensor(0.) tensor(0.0019916897)
tensor(-0.0024998188) tensor(-0.0022729994)
```

It seems that 

https://github.com/pytorch/pytorch/blob/4bb7e148c46167cb2b0beedf4332eb6eae5b03cc/aten/src/ATen/native/LossCTC.cpp#L330

shall be corrected to

``` c++
res = -std::exp(res + nll - lp) * gr;
```

### Versions

PyTorch version: 1.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (x86_64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.18.2
Libc version: N/A

Python version: 3.8.5 (default, Jul 21 2020, 10:48:26)  [Clang 11.0.3 (clang-1103.0.32.62)] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.12.0
[conda] Could not collect

**Comment by vadimkantorov** [comment]
@t-vi if correct, it's impressive this wasn't found by tests...

**Comment by wx-csy** [comment]
The cuda implementation seems to have the same problem ...

https://github.com/pytorch/pytorch/blob/4bb7e148c46167cb2b0beedf4332eb6eae5b03cc/aten/src/ATen/native/cuda/LossCTC.cu#L541

**Comment by wx-csy** [comment]
Aha, I found that the forward and backward steps of ctc_loss are inconsistent...

The forward pass implementation doesn't include a log_softmax layer, but in the backward pass, the gradient is propagated through an extra log_softmax layer. This will not affect if there is already a log_softmax layer before the ctc_loss layer because log_softmax is idempotent; the issue is manifested only when the loss is used alone.

So we have to clarify the definition of the ctc_loss: does it normalize the input by softmax? If so, we should change the forward pass to include the normalization and if not we have to fix the backward pass...

**Comment by vadimkantorov** [comment]
I have a vague recollection that we discussed it with @t-vi somewhere, but I couldn’t find where :(

**Comment by t-vi** [comment]
The gist is that PyTorch projects the gradient to be in the tangent plane of probabilities. This isn't a problem if your inputs do this (ie for most apprications) but I probably would not do it again this way. I have a small patch to change the behaviour but it does seem to have triggered some test failures.

**Comment by Lezcano** [comment]
Note that this is the same issue we face when dealing with any other operation with constrained inputs (e.g. `linalg.eigh` or `linalg.cholesky`). 

In these cases, one has to be careful when computing the adjoints in the backward pass. In particular, the adjoint of a function on a restricted submanifold (isometrically embedded) is the orthogonal projection of the adjoint of the function on the total space. Here's where these projections pop up.

I wrote a slightly mathematical note about this when writing `cholesky_backward`:
https://github.com/pytorch/pytorch/blob/42fefd4403829406427b0e2912d28290d60ef86a/torch/csrc/autograd/FunctionsManual.cpp#L1611-L1620

I do not know enough about this function in particular to comment here, but the only thing I'd like to note is that, if we have a function defined on some subset (submanifold really) of the inputs and this function is differentiable, then the derivative is mathematically defined in a unique way, meaning that there's no room for choices of "whether to project this variable here or not". There's one and only mathematical definition for its derivative.

I can have a look at this one in a few days if you believe that this could be of help.

**Comment by wx-csy** [comment]
> Note that this is the same issue we face when dealing with any other operation with constrained inputs (e.g. `linalg.eigh` or `linalg.cholesky`).
> 
> In these cases, one has to be careful when computing the adjoints in the backward pass. In particular, the adjoint of a function on a restricted submanifold (isometrically embedded) is the orthogonal projection of the adjoint of the function on the total space. Here's where these projections pop up.
> 
> I wrote a slightly mathematical note about this when writing `cholesky_backward`:
> 
> https://github.com/pytorch/pytorch/blob/42fefd4403829406427b0e2912d28290d60ef86a/torch/csrc/autograd/FunctionsManual.cpp#L1611-L1620
> 
> I do not know enough about this function in particular to comment here, but the only thing I'd like to note (at risk of stating the obvious) is that, if we have a function defined on some subset (submanifold really) of the inputs and this function is differentiable, then the derivative is mathematically defined in a unique way. This means that there's no room for choices of "whether to project this variable here or not". There's one and only mathematical definition for its derivative (although there may be many mathematically equivalent implementations with different efficiency and numerical properties, of course).
> 
> I can have a look at this one in a few days if you believe that this could be of help.

Good point. The situation about this specific function is that it can accept inputs outside the probability plane, but the gradients are computed as if it is restricted on the probability plane. The canonical interpretation of CTCLoss asserts that the inputs are probabilities, but I'm not sure whether CTCLoss can be used on unnormalized inputs with some practical meaning. Actually nothing except the probabilistic interpretation prevents it from accepting unnormalized inputs; in particular, since CTCLoss can be written as a multivariate polynomial of its inputs, it is mathematically well-defined over the entire space.

Maybe we should well-specify the domain of this function...

**Comment by Lezcano** [comment]
> Maybe we should well-specify the domain of this function...

I think this is the right approach. Interestingly enough, we have a similar problem in `kl_div`, where the inputs should also be probability measures but we do not want to check that explicitly https://github.com/pytorch/pytorch/issues/80488#issuecomment-1203976179

**Comment by t-vi** [comment]
So I have dropped the projection on the probability gradient and most of the expected tests passed, but not all, I need to investigate those.
